### PR TITLE
Next 13 fixes

### PR DESCRIPTION
**Pull request by jamesdaniels**
Couple of fixes for Next.js v13

* Bump deploy test to Next 13, add an app directory and some additional components
* Set __NEXT_REACT_ROOT when compiling an app targeting React 18
* Don't blindly copy over all .html files, we need to skip those using ISR
* Skip ISR files while building the hosting directory
* Generated html/json can be found in the app directory, not just pages, now
* Removed some unused deps
* Use NPM on Next.js hosting init

TODO

* [x] add comment clarifying we shouldn't go overboard optimizing for app directory yet, as it's unstable
* [x] change log
* [x] cleanup types
* [x] move to semver test for react 18

### CODE REVIEW

**Comment by TheIronDev** [comment]
Please add tests to capture whats changing in `src/frameworks/next/index.ts`

**Comment by jamesdaniels** [comment]
@TheIronDev added into the nextjs deploy test. FYI I'm porting the matrix/canary tests from FirebaseExtended to this repo, WIP here #5187 aiming to have that in good shape early next week. This would catch problems like this.

**Review by inlined — APPROVED** [review]

**Comment by hassaanp** [comment]
when can we expect this to merge?

File: src/frameworks/next/index.ts [hunk 1]
```diff
@@ -176,42 +197,59 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
+    }
 
     const prerenderManifestBuffer = await readFile(
       join(sourceDir, distDir, "prerender-manifest.json")
     );
     const prerenderManifest = JSON.parse(prerenderManifestBuffer.toString());
-    // TODO drop from hosting if revalidate
-    for (const route in prerenderManifest.routes) {
-      if (prerenderManifest.routes[route]) {
+    for (const path in prerenderManifest.routes) {
```
[the remark this answers is not in the document]
Comment (jamesdaniels) [reply src/frameworks/next/index.ts:217]: it's not an iterable, it's an object: `{ "/foo": { ... }, "/bar": { ... }}`

so in is correct, though with the linting rule (to ensure we don't iterate over private) it's a little verbose... perhaps `const path of Object.keys(prerenderManifest)` or `const [path, attr] of Object.entries(prerenderManifest.routes)` would be stylistically preferable?

File: src/frameworks/next/index.ts [hunk 2]
```diff
@@ -176,42 +197,59 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
+      if (prerenderManifest.routes[path]) {
+        // Skip ISR in the deploy to hosting
+        const { initialRevalidateSeconds } = prerenderManifest.routes[path];
+        if (initialRevalidateSeconds) continue;
```
[the remark this answers is not in the document]
Comment (jamesdaniels) [reply src/frameworks/next/index.ts:221]: SG

File: src/frameworks/next/index.ts [hunk 3]
```diff
@@ -176,42 +197,59 @@ export async function ɵcodegenPublicDirectory(sourceDir: string, destDir: strin
...
+
         // / => index.json => index.html => index.html
         // /foo => foo.json => foo.html
-        const parts = route
+        const parts = path
           .split("/")
           .slice(1)
           .filter((it) => !!it);
```
[the remark this answers is not in the document]
Comment (jamesdaniels) [reply src/frameworks/next/index.ts:228]: I was likely overly defensive here, worrying aboth leading/trailing slash, I'll check my assumptions. that said, we're pushing cleanURLs to be true by default now when paired with hosting.source, I can probably simplify this code. I'll do that in a follow on so I unblock Next 13 from working

[the remark this answers is not in the document]
Comment (jamesdaniels) [reply src/frameworks/next/index.ts:233]: Took on as a TODO to explore simplifying

**Comment by LorenzoBloedow** [comment]
When can we expect the next release with this fix? I tried building it myself a few days ago but had a weird bug where it kept
duplicating my whole project, probably did something wrong in the build process.

**Comment by jeremyrajan** [comment]
i tried the 11.15.0 release, doesnt seem to work tho. Still giving me error that its looking for `pages` and not `app`. Anything I am doing wrong?

**Comment by RajendraAVerma** [comment]
> i tried the 11.15.0 release, doesnt seem to work tho. Still giving me error that its looking for `pages` and not `app`. Anything I am doing wrong?

I tried the 11.15.0 and release and hopefully it work 🔥👍 ! 
```
 "dependencies": {
    "eslint": "8.26.0",
    "eslint-config-next": "13.0.1",
    "next": "13.0.1",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  }
```

**Comment by luisorbaiceta** [comment]
> i tried the 11.15.0 release, doesnt seem to work tho. Still giving me error that its looking for `pages` and not `app`. Anything I am doing wrong?

Version 11.15.0 is from two weeks ago and this has been merged 12 hours ago

**Comment by JGSolutions** [comment]
do u guys still use npm build && npm export?

**Comment by jamesdaniels** [comment]
@JGSolutions ATM we call the Node entry point for [Next Build API](https://nextjs.org/docs/deployment#nextjs-build-api) `next/dist/build` and we attempt a `next export`. If the export succeeds we deploy that to the CDN. We then use heuristics to determine if a Cloud Functions backend is needed.

In an upcoming release, we're dropping the attempted `next export` as we've found shortcomings in its assumptions that make it an unreliable early return for building the CDN artifacts. Our own "export" out of the Next Built API is proving just as reliable now that we've built up a good understanding of the manifest files.

**Comment by sanesanyo** [comment]
@jamesdaniels Can you please look at this thread: https://github.com/firebase/firebase-tools/issues/5369 ? We have been struggling with Firebase NextJS deployment issues and so far no one has answered from Firebase side. Given you are working on the development, hopefully you can help us out. Thanks a lot in advance.

**Pull request merged**